Firestore/Source/Util/FSTDispatchQueue.mm

```diff
@@ -21,8 +21,130 @@
+ * queue after the provided delay.
+ * @param queue The FSTDispatchQueue to run the callback on.
+ * @param timerID A FSTTimerID identifying the type of delayed callback this is.
+ * @param delay The delay (seconds) before the callback should be scheduled.
+ * @param callback The callback block to run.
+ * @return The created FSTDelayedCallback instance.
+ */
++ (instancetype)createAndScheduleWithQueue:(FSTDispatchQueue *)queue
+                                   timerID:(FSTTimerID)timerID
+                                     delay:(NSTimeInterval)delay
+                                  callback:(void (^)(void))callback;
+
+/**
+ * Queues the operation to run immediately (if it hasn't already been run or canceled).
+ */
+- (void)skipDelay;
+
+@end
+
+@implementation FSTDelayedCallback
+
+- (instancetype)initWithQueue:(FSTDispatchQueue *)queue
+                      timerID:(FSTTimerID)timerID
+                   targetTime:(NSTimeInterval)targetTime
+                     callback:(void (^)(void))callback {
+  if (self = [super init]) {
+    _queue = queue;
+    _timerID = timerID;
+    _targetTime = targetTime;
+    _callback = callback;
+    _pending = YES;
+  }
+  return self;
+}
+
++ (instancetype)createAndScheduleWithQueue:(FSTDispatchQueue *)queue
+                                   timerID:(FSTTimerID)timerID
+                                     delay:(NSTimeInterval)delay
+                                  callback:(void (^)(void))callback {
+  NSTimeInterval targetTime = [[NSDate date] timeIntervalSince1970] + delay;
+  FSTDelayedCallback *delayedCallback = [[FSTDelayedCallback alloc] initWithQueue:queue
+                                                                          timerID:timerID
+                                                                       targetTime:targetTime
+                                                                         callback:callback];
+  [delayedCallback startWithDelay:delay];
+  return delayedCallback;
+}
+
+/**
+ * Starts the timer. This is called immediately after construction by createAndScheduleWithQueue.
+ */
+- (void)startWithDelay:(NSTimeInterval)delay {
+  dispatch_time_t delayNs = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC));
+  dispatch_after(delayNs, self.queue.queue, ^{
+    [self delayDidElapse];
+  });
+}
+
+- (void)skipDelay {
+  [self.queue dispatchAsyncAllowingSameQueue:^{
+    [self delayDidElapse];
+  }];
+}
+
+- (void)cancel {
+  [self.queue verifyIsCurrentQueue];
+  if (self.isPending) {
+    // PORTING NOTE: There's no way to actually cancel the dispatched callback, but it'll be a no-op
+    // since we set pending to NO.
+    [self clear];
+  }
+}
+
+- (void)delayDidElapse {
+  [self.queue verifyIsCurrentQueue];
+  if (self.isPending) {
+    [self clear];
+    self.callback();
+  }
+}
+
+/**
+ * Marks this delayed callback as no longer pending, and notifies the FSTDispatchQueue that it
+ * should be removed.
+ */
+- (void)clear {
```

Comment:
"clearTimeout" is a pretty descriptive name on the Web. "Clear" dropps this descriptiveness. Maybe 'clearOperation'?

Reply:
Yeah, I struggled to name this. I don't think 'Operation' is particularly suitable because (unlike web) we don't use the term "operation" in FSTDispatchQueue at all.

I opted to invert the `pending` flag into a `done` flag, and rename this `markDone`. Let me know what you think.